docs/rfc/012-record-transaction-model.md

```diff
@@ -0,0 +1,243 @@
+The types of the slots would be fairly low-level,
+i.e. byte arrays (e.g. `u8[128]` for the payload)
+and unsigned integers (e.g. `u64` for the balance),
+because they must have a clear correspondence with the serialized form of records.
+This means that the Leo code may have to do
+its own deserialization of the payload bytes into higher-level Leo values;
+standard serialization/deserialization libraries for Leo types may be provided for this,
+as an independent and more generally useful feature.
+
+It may make sense to have a circuit type for the special `input` variable,
+which includes two slots for the two old records.
+All these circuit types should be explicitly documented,
+and available to the Leo program.
+
+However, we probably want `input` to be read-only,
+i.e. disallow assigning to an old record slot.
+Designating `input` as `const` does not seem right,
+as that designation normally means that it is compiled into the circuit.
+Instead, we could provide read-only access via member function (e.g. `get_payload()`, `get_balance()`),
+but we still have to prohibit assignments to member variables (which is currently allowed on any circuit type).
+As an orthogonal and more generally useful feature,
+we could consider adding public/private access designations to Leo circuit members.
+Another approach is to avoid exposing the member variables,
+and just make the member functions available via an implicit import declaration.
+All of this needs to be thought through more carefully, in the broader context of the Leo language design;
+in any case, it should be clear that this can be made to work in some way,
+and that Leo programs can access the old records through the special `input` variables.
+
+One issue with the special `input` variable is whether it should be treated as a built-in global variable,
+or whether it should be explicitly passed to the entry point functions and to the non-entry-point functions called by them.
+The first approach is more concise, while the second approach is more explicit.
+Note that, in the second approach, we may want to enforce certain restrictions on the use of `input`,
+e.g. we may not want to allow a call `f(input, input)` even if the parameters of `f` both have the same circuit type as `input`.
+That is, even if we make the passing of `input` explicit,
+there is still something special about it;
+in essence, it still needs to be treated essentially like a global variable,
+i.e. in a single-threaded way.
+
+The treatment of output records is less clear at this point.
+As mentioned above, experimentation suggests that currently the output values of `main` are serialized into new records.
+This is not "symmetric" with the treatment of input records.
+It may be preferable to require the Leo code to perform its own serialization of high-level data to output records,
+which would often be the inverse of the deserialization from input records.
+We could consider, for symmetry, to add a special `output` variable,
```

Comment:
For analysis purposes, function outputs tend to work better than implicit globals.

Reply:
I agree in general. I'm concerned about the implicit serialization of the outputs into the new records vs. the explicit serialization of the old records...

I'm wondering if it might make sense to have the entry point functions return records (or portions of them) explicitly, i.e. to require entry point functions to return one of the following types:
```
() // for when no new record is created
Record // for when one new record is created
(Record, Record) // for when two new records are created
```
This way we avoid the implicit serialization and avoid an `output` global.

Reply:
I just pushed some new text about this.